### ISSUE TITLE
IR explosion on SPIR-V switches

### ISSUE THREAD

**Issue by kvark** [gfx-rs/naga#691]
[394.spv.zip](https://github.com/gfx-rs/naga/files/6289197/394.spv.zip)
Flow [flow.Fragment-main.dot.zip](https://github.com/gfx-rs/naga/files/6289200/flow.Fragment-main.dot.zip)
Converting this SPIRV to MSL produces a 300Mb file...

**Comment by kvark** [comment]
It seems to me that this is the same problem as #690, just in the processing of switches.

**Comment by kvark** [comment]
@MatusT It looks like we have a few major issues related to `OpSwitch`:
  1. we are only expecting a case to branch to the *following* case, while SPIRV spec allows it to branch to one other case, wherever it is. So we need to reconstruct this branch tree, and sort the cases before producing the IR.
  2. we aren't stopping the traversal on reaching the continue target. This means we are repeating over and over the same case that we descend into from other cases. That means IR explosion in the end.

For (2), I'm not sure if it requires us to support multiple "stop nodes" (i.e. `&[NodeIndex]` instead of `Some(NodeIndex)`), or some classification logic to find and mark those edges first, preventing the traversal this way.

**Comment by MatusT** [comment]
> 1. we are only expecting a case to branch to the _following_ case, while SPIRV spec allows it to branch to one other case, wherever it is. So we need to reconstruct this branch tree, and sort the cases before producing the IR.

Doesn't this contradict the spec?
```
if Target T1 branches to Target T2, or if Target T1 branches to the Default and the Default branches to Target T2, then T1 must immediately precede T2 in the list of the OpSwitch Target operands
```

> For (2), I'm not sure if it requires us to support multiple "stop nodes" (i.e. `&[NodeIndex]` instead of `Some(NodeIndex)`), or some classification logic to find and mark those edges first, preventing the traversal this way.

So far, it seems that any block should be traversed only once AND there is a need only for one stop. I would start with marking that would happen only in debug mode. That way, we can find where it fails without introducing runtime overhead. If we find a case that would require more "stop nodes" then we can add that. But I think that is unlikely as even Tint doesn't need them.

**Comment by kvark** [comment]
Ok, thank you! I assuming this is also waiting for #688 then.

**Comment by kvark** [comment]
I think this is fixed by #688
